packages/flutter_tools/lib/src/android/migrations/min_sdk_version_migration.dart

```diff
@@ -0,0 +1,47 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:meta/meta.dart';
+
+import '../../base/project_migrator.dart';
+import '../../project.dart';
+
+@visibleForTesting
+const String minSdk16 = 'minSdkVersion 16';
+@visibleForTesting
+const String minSdk17 = 'minSdkVersion 17';
+@visibleForTesting
+const String minSdk18 = 'minSdkVersion 18';
+@visibleForTesting
+const String flutterMinSdk = 'minSdkVersion flutter.minSdkVersion';
+@visibleForTesting
+const String appGradleNotFoundWarning = 'Module level build.gradle file not found, skipping minSdkVersion migration.';
+
+class MinSdkVersionMigration extends ProjectMigrator {
+  MinSdkVersionMigration(
+      AndroidProject project,
+      super.logger,
+  ) : _project = project;
+
+
+  final AndroidProject _project;
+
+  @override
+  void migrate() {
+    // Skip if we cannot find the app level build.gradle file.
+    if (!_project.appGradleFile.existsSync()) {
```

Comment:
This pattern:

if (file exists) {
  open file
}

or

if (!file exists) {
  return
}
open file

is inherently racy and should be avoided (file can be deleted/created between the exists check and opening). 

That is, processFileLines should return an error (or throw an exception) in the case that the file doesn't exist when it tries to open the file.

Reply:
Good call! I replaced all instances where I was relying on an `existsSync` result to instead be wrapped in a `try/catch FileSystemException`. 

There are a lot more that exist outside of this PR (in project.dart, and other migrators). Do you think it's worth creating an issue for the general task of replacing other instances of the anti-pattern? I can create it if so.

Reply:
Yes, please create an issue- I don't think this is urgent but it would be good to get rid of over time.

Reply:
https://github.com/flutter/flutter/issues/129746